clippy_lints/src/replace_consts.rs

```diff
@@ -0,0 +1,102 @@
+use rustc::lint::*;
+use rustc::hir;
+use rustc::hir::def::Def;
+use utils::{match_def_path, span_lint_and_sugg};
+
+/// **What it does:** Checks for usage of `ATOMIC_X_INIT`, `ONCE_INIT`, and
+/// `uX/iX::MIN/MAX`.
+///
+/// **Why is this bad?** `const fn`s exist
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// static FOO: AtomicIsize = ATOMIC_ISIZE_INIT;
+/// ```
+///
+/// Could be written:
+///
+/// ```rust
+/// static FOO: AtomicIsize = AtomicIsize::new(0);
+/// ```
+declare_lint! {
+    pub REPLACE_CONSTS,
+    Allow,
+    "Lint usages of standard library `const`s that could be replaced by `const fn`s"
+}
+
+pub struct ReplaceConsts;
+
+impl LintPass for ReplaceConsts {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(REPLACE_CONSTS)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ReplaceConsts {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx hir::Expr) {
+        if_chain! {
+            if let hir::ExprPath(ref qp) = expr.node;
+            if let Def::Const(def_id) = cx.tables.qpath_def(qp, expr.hir_id);
+            then {
+                for &(const_path, repl_snip) in REPLACEMENTS {
+                    if match_def_path(cx.tcx, def_id, const_path) {
+                        span_lint_and_sugg(
+                            cx,
+                            REPLACE_CONSTS,
+                            expr.span,
+                            &format!("using `{}`", const_path.last().expect("empty path")),
+                            "try this",
+                            repl_snip.to_string(),
+                        );
+                        return;
+                    }
+                }
+            }
+        }
+    }
+}
+
+const REPLACEMENTS: &[(&[&str], &str)] = &[
+    // Once
+    (&["core", "sync",  "ONCE_INIT"], "Once::new()"),
+    // Atomic
+    (&["core", "sync", "atomic", "ATOMIC_BOOL_INIT"],  "AtomicBool::new(false)"),
+    (&["core", "sync", "atomic", "ATOMIC_ISIZE_INIT"], "AtomicIsize::new(0)"),
+    (&["core", "sync", "atomic", "ATOMIC_I8_INIT"],    "AtomicI8::new(0)"),
+    (&["core", "sync", "atomic", "ATOMIC_I16_INIT"],   "AtomicI16::new(0)"),
+    (&["core", "sync", "atomic", "ATOMIC_I32_INIT"],   "AtomicI32::new(0)"),
+    (&["core", "sync", "atomic", "ATOMIC_I64_INIT"],   "AtomicI64::new(0)"),
+    (&["core", "sync", "atomic", "ATOMIC_USIZE_INIT"], "AtomicUsize::new(0)"),
+    (&["core", "sync", "atomic", "ATOMIC_U8_INIT"],    "AtomicU8::new(0)"),
+    (&["core", "sync", "atomic", "ATOMIC_U16_INIT"],   "AtomicU16::new(0)"),
+    (&["core", "sync", "atomic", "ATOMIC_U32_INIT"],   "AtomicU32::new(0)"),
+    (&["core", "sync", "atomic", "ATOMIC_U64_INIT"],   "AtomicU64::new(0)"),
+    // Min
+    (&["core", "isize", "MIN"], "isize::min_value()"),
```

Comment:
`MIN`/`MAX` are not bad at all, I'd rather recommend the opposite replacement.

Reply:
I'll add a config option for enabling `MIN`/`MAX` on this lint. The opposite replacement might be better suited for a different lint though.

Reply:
@petrochenkov why? the `MIN` is in `std::isize::MIN` and thus requires the full path or an import. `min_value` exists on the type

Reply:
That's actually the main reason I recommend these; fewer paths to import.

Reply:
I sort of assumed that it'd make sense to long-term push people toward the `const fn` versions over the constants. For the `INIT` structs, they simply aren't at clear as using `new`, and for the `MIN/MAX` structs, the types themselves are already in scope.

Reply:
>the `MIN` is in `std::isize::MIN` and thus requires the full path or an import.

Oh. I thought they are [associated constants](https://github.com/rust-lang/rust/pull/28656) already.
Someone should resurrect that PR.

>I sort of assumed that it'd make sense to long-term push people toward the const fn versions over the constants.

Why? Constants exist in the language to represent... constants. `MIN`/`MAX` are constants, why would you avoid a feature specifically designed to represent them (except for the full path issue which is a fixable library deficiency). `min_value()` is longer and has that extraneous call.
(I agree about `INIT` atomics though.)

Reply:
^^ This was a concern about constants and const functions in general.
All the listed constants have specific reasons to avoid them (until https://github.com/rust-lang/rust/pull/28656 is redone), so this PR is okay.

Reply:
I totally agree with you @petrochenkov in terms of making them associated constants. When you do end up redoing that PR, you should include the float constants too, as those don't have associated const fns.

Reply:
So, no config option for `MIN`/`MAX` then?

Reply:
Nah, leave it out, we'll adjust the lint when it makes sense to behave differently on stable